src/borg/helpers.py

```diff
@@ -728,16 +728,16 @@ class Location:
     # the archive names. Thus, we must not accept "/" in archive names.
     ssh_re = re.compile(r'(?P<proto>ssh)://(?:(?P<user>[^@]+)@)?'
                         r'(?P<host>[^:/#]+)(?::(?P<port>\d+))?'
-                        r'(?P<path>[^:]+)(?:::(?P<archive>[^/]+))?$')
+                        r'(?P<path>[^:]+)(?:::(?P<archive>[^/\s]([^/\s]| )*[^/\s]))?$')
     file_re = re.compile(r'(?P<proto>file)://'
-                         r'(?P<path>[^:]+)(?:::(?P<archive>[^/]+))?$')
+                         r'(?P<path>[^:]+)(?:::(?P<archive>[^/\s]([^/\s]| )*[^/\s]))?$')
     scp_re = re.compile(r'((?:(?P<user>[^@]+)@)?(?P<host>[^:/]+):)?'
-                        r'(?P<path>[^:]+)(?:::(?P<archive>[^/]+))?$')
+                        r'(?P<path>[^:]+)(?:::(?P<archive>[^/\s]([^/\s]| )*[^/\s]))?$')
     # get the repo from BORG_RE env and the optional archive from param.
     # if the syntax requires giving REPOSITORY (see "borg mount"),
     # use "::" to let it use the env var.
     # if REPOSITORY argument is optional, it'll automatically use the env.
-    env_re = re.compile(r'(?:::(?P<archive>[^/]+)?)?$')
+    env_re = re.compile(r'(?:::(?P<archive>[^/\s]([^/\s]| )*[^/\s])?)?$')
```

Comment:
This is not just related to you change, but more a general thing:
this code / these regexes is somehow way to complex already (and it's not getting better by making the regexes even more complex).
So, I'ld rather like to simplify / deduplicate it somehow.